pkg/controller/deliverable/reconciler.go

```diff
@@ -0,0 +1,184 @@
+	if err != nil {
+		switch typedErr := err.(type) {
+		case realizer.GetDeliveryClusterTemplateError:
+			r.conditionManager.AddPositive(TemplateObjectRetrievalFailureCondition(typedErr))
+		case realizer.StampError:
+			r.conditionManager.AddPositive(TemplateStampFailureCondition(typedErr))
+		case realizer.ApplyStampedObjectError:
+			r.conditionManager.AddPositive(TemplateRejectedByAPIServerCondition(typedErr))
+		case realizer.RetrieveOutputError:
+			r.conditionManager.AddPositive(MissingValueAtPathCondition(typedErr.ComponentName(), typedErr.JsonPathExpression()))
+			err = nil
+		default:
+			r.conditionManager.AddPositive(UnknownResourceErrorCondition(typedErr))
+		}
+
+		return r.completeReconciliation(reconcileCtx, deliverable, err)
+	}
+
+	r.conditionManager.AddPositive(ResourcesSubmittedCondition())
+
+	return r.completeReconciliation(reconcileCtx, deliverable, nil)
+}
+
+func (r *Reconciler) completeReconciliation(ctx context.Context, deliverable *v1alpha1.Deliverable, err error) (ctrl.Result, error) {
+	logger := logr.FromContext(ctx)
+
+	var changed bool
+	deliverable.Status.Conditions, changed = r.conditionManager.Finalize()
+
+	var updateErr error
+	if changed || (deliverable.Status.ObservedGeneration != deliverable.Generation) {
+		deliverable.Status.ObservedGeneration = deliverable.Generation
+		updateErr = r.repo.StatusUpdate(deliverable)
+		if updateErr != nil {
+			logger.Error(updateErr, "update error")
+			if err == nil {
+				logger.Info("finished")
+				return ctrl.Result{}, fmt.Errorf("update deliverable status: %w", updateErr)
+			}
+		}
+	}
+
+	logger.Info("finished")
+
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	if !r.conditionManager.IsSuccessful() { // TODO: Discuss rename to IsReady
+		return ctrl.Result{}, fmt.Errorf("deliverable not ready")
+	}
+
+	return ctrl.Result{RequeueAfter: reconcileInterval}, nil
+}
+
+func (r *Reconciler) checkDeliveryReadiness(delivery *v1alpha1.ClusterDelivery) error {
+	readyCondition := getDeliveryReadyCondition(delivery)
+	if readyCondition.Status == "True" {
+		return nil
+	}
+	return fmt.Errorf("delivery is not in ready condition")
+}
+
+func getDeliveryReadyCondition(delivery *v1alpha1.ClusterDelivery) metav1.Condition {
+	for _, condition := range delivery.Status.Conditions {
+		if condition.Type == "Ready" {
+			return condition
+		}
+	}
+	return metav1.Condition{}
+}
+
+func (r *Reconciler) getDeliveriesForDeliverable(deliverable *v1alpha1.Deliverable) (*v1alpha1.ClusterDelivery, error) {
+	if len(deliverable.Labels) == 0 {
+		r.conditionManager.AddPositive(DeliverableMissingLabelsCondition())
+		return nil, fmt.Errorf("deliverable is missing required labels")
+	}
+
+	deliveries, err := r.repo.GetDeliveriesForDeliverable(deliverable)
+	if err != nil || len(deliveries) == 0 {
+		r.conditionManager.AddPositive(DeliveryNotFoundCondition(deliverable.Labels))
+
+		if err != nil {
```

Comment:
wdyt of flattening this all out?

e.g.:

```go
          if err != nil {
                  r.conditionManager.AddPositive(DeliveryNotFoundCondition(deliverable.Labels))
                  return nil, fmt.Errorf("get delivery by label: %w", err)
          }

          if len(deliveries) == 0 {
                  r.conditionManager.AddPositive(DeliveryNotFoundCondition(deliverable.Labels))
                  return nil, fmt.Errorf("no delivery found where full selector is satisfied by labels: %v", deliverable.Labels)
          }

          if len(deliveries) > 1 {
                  r.conditionManager.AddPositive(TooManyDeliveryMatchesCondition())
                  return nil, fmt.Errorf("too many deliveries match the deliverable selector label")
          }
```

I think the less nesting the better as then it becomes more apparent what the necessary conditions to be fulfilled are, but that might be just personal taste